Gradio UI

freddyaboulton opened the pull request:
Converting the UI from streamlit to gradio.

The approach I went with was to create a `widgets` module that implements each of the widgets in separate files. The widgets (along with the rest of the ui) are rendered is defined in `app.main`.

Each widget inherits from a `Widget` base clase and implements a a few abstract methods. I think having the abstract base class will make it easier to implement new widgets in the future. 

There are a couple of gradio-side enhancements I want to make to the dataframe component to improve how some of the datasets are displayed. These are:

https://github.com/gradio-app/gradio/issues/2667
https://github.com/gradio-app/gradio/issues/2666
https://github.com/gradio-app/gradio/issues/2688

The core functionality should be all there though. You will get errors in the UI if you select a dataset/config/feature combo that has not been cached yet. Reading from the cache will take a long time for npmi for some datasets so we should implement a better cache format in a subsequent pr.

freddyaboulton commented:
@meg-huggingface I think #44 made it so that the Label distribution never shows up?

meg-huggingface commented on run.sh:
```diff
@@ -5,12 +5,12 @@ python3 run_data_measurements.py --dataset="hate_speech18" --config="default" --
 python3 run_data_measurements.py --dataset="hate_speech_offensive" --config="default" --split="train" --label_field="label" --feature="tweet"
```
Why not add --overwrite previous to all of these?

freddyaboulton replied:
Can do!

meg-huggingface commented on app.py:
```diff
@@ -184,28 +176,102 @@ def main():
     pull_cache_from_hub = arguments.pull_cache_from_hub
```
Note that `live` and `pull_cache_from_hub` are defined here -- these need to be used later (now on line 211; I made a separate comment).

freddyaboulton replied:
Done!

meg-huggingface commented on app.py:
```diff
@@ -184,28 +176,102 @@ def main():
+                                     ("npmi", dstats.load_or_prepare_npmi),
+                                     ("zipf", dstats.load_or_prepare_zipf)]
+                dstats = load_or_prepare_widgets(dstats, load_prepare_list, show_perplexities=False,
+                                                 live=True, pull_cache_from_hub=False)
```
```suggestion
                                                 live=live, pull_cache_from_hub=pull_cache_from_hub)
```

meg-huggingface commented on app.py:
```diff
@@ -184,28 +176,102 @@ def main():
+        widget_list = [widgets.DatasetDescription(DATASET_NAME_TO_DICT),
+                       widgets.GeneralStats(),
+                       widgets.LabelDistribution(),
+                       widgets.TextLengths(),
```
After this is called, I get the following error and it crashes:



2022-11-29 15:13:41.234 python[43481:9115201] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
        0   CoreFoundation                      0x00007fff3460cdd7 __exceptionPreprocess + 250
        1   libobjc.A.dylib                     0x00007fff6d06d5bf objc_exception_throw + 48
        2   CoreFoundation                      0x00007fff346355bc -[NSException raise] + 9
        3   AppKit                              0x00007fff3183296c -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 310
        4   AppKit                              0x00007fff3181a3d2 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1416
        5   AppKit                              0x00007fff31819e43 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
        6   _macosx.cpython-39-darwin.so        0x0000000186f845ac -[Window initWithContentRect:styleMask:backing:defer:withManager:] + 76
        7   _macosx.cpython-39-darwin.so        0x0000000186f87a49 FigureManager_init + 281
        8   python                              0x0000000106c4bbac wrap_init + 12
        9   python                              0x0000000106bcf8ef wrapperdescr_call + 911
        10  python                              0x0000000106cdf147 _PyEval_EvalFrameDefault + 52951
        11  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        12  python                              0x0000000106c4b50a slot_tp_init + 346
        13  python                              0x0000000106c50ce0 type_call + 272
        14  python                              0x0000000106cdf147 _PyEval_EvalFrameDefault + 52951
        15  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        16  python                              0x0000000106bc79ae method_vectorcall + 158
        17  python                              0x0000000106cdbd9e _PyEval_EvalFrameDefault + 39726
        18  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        19  python                              0x0000000106bc79ae method_vectorcall + 158
        20  python                              0x0000000106ce062c _PyEval_EvalFrameDefault + 58300
        21  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        22  python                              0x0000000106ce062c _PyEval_EvalFrameDefault + 58300
        23  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        24  python                              0x0000000106ce062c _PyEval_EvalFrameDefault + 58300
        25  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        26  python                              0x0000000106cd9098 _PyEval_EvalFrameDefault + 28200
        27  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        28  python                              0x0000000106cd913e _PyEval_EvalFrameDefault + 28366
        29  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        30  python                              0x0000000106cda8dd _PyEval_EvalFrameDefault + 34413
        31  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        32  python                              0x0000000106bc79ae method_vectorcall + 158
        33  python                              0x0000000106cd9098 _PyEval_EvalFrameDefault + 28200
        34  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        35  python                              0x0000000106cd9098 _PyEval_EvalFrameDefault + 28200
        36  python                              0x0000000106bc2be2 _PyFunction_Vectorcall + 1234
        37  python                              0x0000000106d17589 _PyObject_VectorcallTstate.4590 + 73
        38  python                              0x0000000106d17471 context_run + 81
        39  python                              0x0000000106c204d1 cfunction_vectorcall_FASTCALL_KEYWORDS + 97
        40  python                              0x0000000106cddb21 _PyEval_EvalFrameDefault + 47281
        41  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        42  python                              0x0000000106cda8dd _PyEval_EvalFrameDefault + 34413
        43  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        44  python                              0x0000000106cda8dd _PyEval_EvalFrameDefault + 34413
        45  python                              0x0000000106bc28b4 _PyFunction_Vectorcall + 420
        46  python                              0x0000000106bc7aef method_vectorcall + 479
        47  python                              0x0000000106dc222a t_bootstrap + 186
        48  python                              0x0000000106d5f804 pythread_wrapper + 36
        49  libsystem_pthread.dylib             0x00007fff6e41a109 _pthread_start + 148
        50  libsystem_pthread.dylib             0x00007fff6e415b8b thread_start + 15
)
libc++abi.dylib: terminating with uncaught exception of type NSException
zsh: abort      python app.py

freddyaboulton replied:
I got that during development because of the use of seaborn (not thread safe). I do not get that locally. Could it be because I recached all of the results?

meg-huggingface commented on app.py:
```diff
@@ -184,28 +176,102 @@ def main():
     pull_cache_from_hub = arguments.pull_cache_from_hub
 
     # Initialize the interface and grab the UI-provided arguments
```
I prefer to have the `main` function serve as a basic entry point for the program, calling to other functions. Each function should be modularized so that each one does one task, rather than having all the code together in one long function. I'm not sure whether this is a kind of `official` practice, but it's definitely what I'm used to and what I'm used to seeing. It helps with debugging and reusing functionality.

freddyaboulton replied:
Good point - most of the gradio demo creation got delegated to `create_demo` so main just parses the command line args and calls `create_demo` now

meg-huggingface commented on app.py:
```diff
@@ -184,28 +176,102 @@ def main():
-    # Write out on the UI what we have.
-    display_measurements(dataset_args, display_list, loaded_dstats,
-                         show_perplexities)
+    with gr.Blocks() as demo:
```
Could all of the logic under this be its own function? It could be called `initialize_demo` or something like this.

freddyaboulton replied:
Yep good point - moved to a `create_demo` function!

meg-huggingface replied:
LGTM.

meg-huggingface commented on widgets/general_stats.py:
```diff
@@ -0,0 +1,60 @@
+            self.general_stats_duplicates,
+        ]
+
+    def add_events(self, state: gr.State):
```
Not entirely sure what this is doing and why it's there, but trusting that it's important!

freddyaboulton replied:
In general, lets us define events that belong to a widget. This widget doesn't have any which is why it's just a `pass`. We need to implement it since I made it an abstract method of the parent class so that future devs would not forget to implement if needed.

meg-huggingface replied:
Gotcha, makes sense!

meg-huggingface commented on widgets/npmi.py:
```diff
@@ -0,0 +1,149 @@
+            s.index.name = "word"
+            s = s.reset_index()
+            bias_col = [col for col in s.columns if col != "word"]
+            # count_cols = s.filter(like="count").columns
```
Can delete.

meg-huggingface commented on widgets/widget_base.py:
```diff
@@ -0,0 +1,24 @@
+from data_measurements.dataset_statistics import DatasetStatisticsCacheClass as dmt_cls
+
+
+class Widget(ABC):
```
Cool!!

meg-huggingface commented:
> @meg-huggingface I think #44 made it so that the Label distribution never shows up?

Hmm I'm not seeing that on the main branch. Is it possible you're noticing labels not showing up for datasets that don't have labels?

freddyaboulton commented:
> Hmm I'm not seeing that on the main branch. Is it possible you're noticing labels not showing up for datasets that don't have labels?

I'm not seeing any labels for `hate_speech`, `hate_speech_offensive`, `imdb`, `squad` on main for default settings. I thought at least some of them did, like `hate_speech`?  

<img width="1475" alt="Screen Shot 2022-11-30 at 8 56 43 AM" src="https://user-images.githubusercontent.com/41651716/204820398-08f458fe-799f-4d34-b10f-16cb7d37c534.png">

<img width="1470" alt="Screen Shot 2022-11-30 at 8 53 02 AM" src="https://user-images.githubusercontent.com/41651716/204820382-5693c06c-7e85-4e98-a9ed-0d7974f35ada.png">

<img width="1471" alt="image" src="https://user-images.githubusercontent.com/41651716/204821095-615cb91c-9a57-4f6c-bfbe-f40939583cec.png">

<img width="1460" alt="image" src="https://user-images.githubusercontent.com/41651716/204821541-35950b67-03dc-475b-ba80-45fafc2f8a4a.png">

freddyaboulton commented on app.py:
```diff
@@ -55,38 +55,29 @@ def get_widgets(dstats):
+    #                 ("npmi", gr_utils.npmi_widget),
+    #                 ("zipf", gr_utils.expander_zipf)]
 
     return load_prepare_list, display_list
```
I refactored this function. It now returns a tuple of 

1.  the ui widget list
2. a function that returns the load_or_prepare functions given a DatasetStatisticsCacheClass

The reason for that is that the widget list is static and is needed when the ui is first created but the load_or_prepare functions depend on a DatasetStatisticsCacheClass which is not created until a user selects dataset/config/split values.

meg-huggingface replied:
Sounds great!

freddyaboulton commented on run_data_measurements.py:
```diff
@@ -31,7 +31,7 @@ def load_or_prepare_widgets(ds_args, show_embeddings=False,
 
     """
-    dstats = dataset_statistics.DatasetStatisticsCacheClass(**ds_args,
+    dstats = dataset_statistics.DatasetStatisticsCacheClass(**ds_args, save=True,
```
I believe I added this when I was trying to figure out why some data was not cached. Will delete.

freddyaboulton commented:
@meg-huggingface Thank you for the review! I believe I addressed all of your comments so far. Let's work together to make sure you can run this + the cache.

meg-huggingface reviewed (commented):
This is looking great!
There are a few critical UI things I see going awry:

(1) The "sidebar" now takes up half the page, making all of the results smaller and harder to see (check out in particular the Text Lengths figure). Can this be changed? The sidebar should ideally be 1/4 (or less) of the page, really working as something on the side.

(2) Scrollbars for tables are no longer there -- we'd like to have each widget stay within some limited size, and then scrolled.  You can see this in the "Dataset Description" table, which also has rows of different heights, in the text lengths, and in nPMI.  Is it possible to shorten these tables and allow them to be scrolled instead?  Don't worry too much about the table with all the rows being displayed at different heights in the "Dataset Description" widget, because that table should be changed to the "Dataset Preview" table we see for datasets on the Hub anyway. But for the nPMI table and for the text lengths table, they should be scrollable so as not to all show at once. Is it possible?

(3) The npmi widget columns can no longer can be sorted. This is a critical functionality -- can we get it back?

(4) Minor: The "proportion" column of the table in the "General Text Statistics" widget should be rounded to 4 decimal places.

Thanks!

meg-huggingface commented:
> @meg-huggingface Thank you for the review! I believe I addressed all of your comments so far. Let's work together to make sure you can run this!

Looking good! Since I still couldn't run it locally, I put in a patch: https://github.com/huggingface/data-measurements-tool/pull/50

Once that's in, the decimal places on everything still need to be rounded -- is it possible?

Thanks!

freddyaboulton commented:
Thank you for the fix in #50!  Just pushed up the rounding too 😊

meg-huggingface reviewed (approved):
At this point the code works and has the basic functionality, but there are remaining TBDs. In particular, getting the nPMI stuff to sort. =)

meg-huggingface commented:
!! SWITCH TO GRADIO UI !! WOOWOO!!
